File: ops/worker/gonic_client.py
"""
Gonic API Client
Handles library scanning and index management
"""
import httpx
from typing import Optional, Dict, Any


class GonicClient:
    """Client for Gonic Subsonic-compatible server"""

    def __init__(self, base_url: str, username: str = "admin", password: str = "admin"):
        """
        Initialize Gonic client

        Args:
            base_url: Base URL of Gonic server (e.g., http://gonic:4747)
            username: Admin username (default from Gonic setup)
            password: Admin password (default from Gonic setup)
        """
        self.base_url = base_url.rstrip('/')
        self.username = username
        self.password = password
        self.client = httpx.AsyncClient(timeout=60.0)

    async def close(self):
        """Close HTTP client"""
        await self.client.aclose()

    async def trigger_scan(self) -> bool:
        """
        Trigger a full library scan

        Returns:
            True if scan was triggered successfully
        """
        try:
            # Gonic Subsonic API endpoint for scanning
            response = await self.client.get(
                f"{self.base_url}/rest/startScan",
                params={
                    "u": self.username,
                    "p": self.password,
                    "v": "1.16.1",
                    "c": "netrunner",
                    "f": "json"
                }
            )

            if response.status_code == 200:
                data = response.json()
                # Check for Subsonic API success
                if "subsonic-response" in data:
                    status = data["subsonic-response"].get("status")
                    return status == "ok"

            return False

        except Exception as e:
            print(f"Error triggering Gonic scan: {e}")
            return False

    async def get_scan_status(self) -> Dict[str, Any]:
        """
        Get current scan status

        Returns:
            Dictionary with scan status information
        """
        try:
            response = await self.client.get(
                f"{self.base_url}/rest/getScanStatus",
                params={
                    "u": self.username,
                    "p": self.password,
                    "v": "1.16.1",
                    "c": "netrunner",
                    "f": "json"
                }
            )

            if response.status_code == 200:
                data = response.json()
                if "subsonic-response" in data:
                    scan_status = data["subsonic-response"].get("scanStatus", {})
                    return {
                        "scanning": scan_status.get("scanning", False),
                        "count": scan_status.get("count", 0)
                    }

            return {"scanning": False, "count": 0}

        except Exception as e:
            print(f"Error getting Gonic scan status: {e}")
            return {"scanning": False, "count": 0}

    async def wait_for_scan_completion(self, check_interval: float = 5.0, timeout: Optional[float] = 300.0):
        """
        Wait for current scan to complete

        Args:
            check_interval: Seconds between status checks
            timeout: Maximum wait time in seconds (None for no timeout)

        Raises:
            TimeoutError: If timeout is reached while scan is still running
        """
        import asyncio

        start_time = asyncio.get_event_loop().time()

        while True:
            status = await self.get_scan_status()

            if not status["scanning"]:
                return

            if timeout is not None:
                elapsed = asyncio.get_event_loop().time() - start_time
                if elapsed >= timeout:
                    raise TimeoutError(f"Scan timeout after {timeout}s")

            await asyncio.sleep(check_interval)

    async def get_music_folders(self) -> list:
        """
        Get configured music folders

        Returns:
            List of music folder dictionaries
        """
        try:
            response = await self.client.get(
                f"{self.base_url}/rest/getMusicFolders",
                params={
                    "u": self.username,
                    "p": self.password,
                    "v": "1.16.1",
                    "c": "netrunner",
                    "f": "json"
                }
            )

            if response.status_code == 200:
                data = response.json()
                if "subsonic-response" in data:
                    folders = data["subsonic-response"].get("musicFolders", {}).get("musicFolder", [])
                    return folders if isinstance(folders, list) else [folders]

            return []

        except Exception as e:
            print(f"Error getting music folders: {e}")
            return []

    async def get_library_stats(self) -> Dict[str, Any]:
        """
        Get library statistics

        Returns:
            Dictionary with artist count, album count, song count
        """
        try:
            # Get artists
            artists_response = await self.client.get(
                f"{self.base_url}/rest/getArtists",
                params={
                    "u": self.username,
                    "p": self.password,
                    "v": "1.16.1",
                    "c": "netrunner",
                    "f": "json"
                }
            )

            artist_count = 0
            album_count = 0

            if artists_response.status_code == 200:
                data = artists_response.json()
                if "subsonic-response" in data:
                    artists = data["subsonic-response"].get("artists", {})
                    index_list = artists.get("index", [])
                    if not isinstance(index_list, list):
                        index_list = [index_list]

                    for index in index_list:
                        artist_list = index.get("artist", [])
                        if not isinstance(artist_list, list):
                            artist_list = [artist_list]
                        artist_count += len(artist_list)
                        # Sum album counts
                        for artist in artist_list:
                            album_count += artist.get("albumCount", 0)

            return {
                "artist_count": artist_count,
                "album_count": album_count
            }

        except Exception as e:
            print(f"Error getting library stats: {e}")
            return {
                "artist_count": 0,
                "album_count": 0
            }

    async def health_check(self) -> bool:
        """Check if Gonic is healthy and reachable"""
        try:
            response = await self.client.get(
                f"{self.base_url}/rest/ping",
                params={
                    "u": self.username,
                    "p": self.password,
                    "v": "1.16.1",
                    "c": "netrunner",
                    "f": "json"
                }
            )

            if response.status_code == 200:
                data = response.json()
                if "subsonic-response" in data:
                    return data["subsonic-response"].get("status") == "ok"

            return False

        except Exception as e:
            print(f"Gonic health check failed: {e}")
            return False

File: backend/internal/services/gonic_client.go
package services

import (
	"context"
	"crypto/md5"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

// GonicClient handles interaction with the Gonic Subsonic-compatible server
type GonicClient struct {
	baseURL  string
	username string
	password string
	client   *http.Client
}

// NewGonicClient creates a new Gonic client
func NewGonicClient(baseURL, username, password string, httpClient *http.Client) *GonicClient {
	if httpClient == nil {
		httpClient = &http.Client{Timeout: 60 * time.Second}
	}
	return &GonicClient{
		baseURL:  fmt.Sprintf("%s/rest", baseURL),
		username: username,
		password: password,
		client:   httpClient,
	}
}

// TriggerScan triggers a full library scan in Gonic
func (c *GonicClient) TriggerScan() (bool, error) {
	var resp struct {
		SubsonicResponse struct {
			Status string `json:"status"`
		} `json:"subsonic-response"`
	}

	err := c.doRequest("startScan", nil, &resp)
	if err != nil {
		return false, err
	}

	return resp.SubsonicResponse.Status == "ok", nil
}

// GetScanStatus retrieves the current scan status
func (c *GonicClient) GetScanStatus() (map[string]interface{}, error) {
	var resp struct {
		SubsonicResponse struct {
			Status     string `json:"status"`
			ScanStatus struct {
				Scanning bool `json:"scanning"`
				Count    int  `json:"count"`
			} `json:"scanStatus"`
		} `json:"subsonic-response"`
	}

	err := c.doRequest("getScanStatus", nil, &resp)
	if err != nil {
		return nil, err
	}

	return map[string]interface{}{
		"scanning": resp.SubsonicResponse.ScanStatus.Scanning,
		"count":    resp.SubsonicResponse.ScanStatus.Count,
	}, nil
}

// GetLibraryStats retrieves library statistics from Gonic
func (c *GonicClient) GetLibraryStats() (map[string]int, error) {
	var resp struct {
		SubsonicResponse struct {
			Artists struct {
				Index []struct {
					Artist []struct {
						AlbumCount int `json:"albumCount"`
					} `json:"artist"`
				} `json:"index"`
			} `json:"artists"`
		} `json:"subsonic-response"`
	}

	err := c.doRequest("getArtists", nil, &resp)
	if err != nil {
		return nil, err
	}

	artistCount := 0
	albumCount := 0
	for _, idx := range resp.SubsonicResponse.Artists.Index {
		artistCount += len(idx.Artist)
		for _, artist := range idx.Artist {
			albumCount += artist.AlbumCount
		}
	}

	return map[string]int{
		"artist_count": artistCount,
		"album_count":  albumCount,
	}, nil
}

// GonicSong represents a track in Gonic
type GonicSong struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Artist string `json:"artist"`
	Album  string `json:"album"`
	Path   string `json:"path"`
}

// Search3 searches for tracks, albums or artists
func (c *GonicClient) Search3(query string) ([]GonicSong, error) {
	params := url.Values{}
	params.Add("query", query)
	params.Add("songCount", "20")

	var resp struct {
		SubsonicResponse struct {
			Status        string `json:"status"`
			SearchResult3 struct {
				Song []GonicSong `json:"song"`
			} `json:"searchResult3"`
		} `json:"subsonic-response"`
	}

	err := c.doRequest("search3", params, &resp)
	if err != nil {
		return nil, err
	}

	return resp.SubsonicResponse.SearchResult3.Song, nil
}

// GetSong retrieves details for a specific song
func (c *GonicClient) GetSong(id string) (*GonicSong, error) {
	params := url.Values{}
	params.Add("id", id)

	var resp struct {
		SubsonicResponse struct {
			Status string    `json:"status"`
			Song   GonicSong `json:"song"`
		} `json:"subsonic-response"`
	}

	err := c.doRequest("getSong", params, &resp)
	if err != nil {
		return nil, err
	}

	if resp.SubsonicResponse.Status != "ok" {
		return nil, fmt.Errorf("subsonic error: %s", resp.SubsonicResponse.Status)
	}

	return &resp.SubsonicResponse.Song, nil
}

func (c *GonicClient) doRequest(endpoint string, params url.Values, target interface{}) error {
	if params == nil {
		params = url.Values{}
	}

	// Subsonic token-based auth (keeps password out of URL query params / server logs)
	s, err := salt()
	if err != nil {
		return err
	}
	params.Add("u", c.username)
	params.Add("t", tokenFromPassword(c.password, s))
	params.Add("s", s)
	params.Add("v", "1.16.1")
	params.Add("c", "netrunner")
	params.Add("f", "json")

	fullURL := fmt.Sprintf("%s/%s?%s", c.baseURL, endpoint, params.Encode())

	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, fullURL, nil)
	if err != nil {
		return err
	}

	resp, err := c.client.Do(req)
	if err != nil {
		return err
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("gonic api error: %s", resp.Status)
	}

	return json.NewDecoder(resp.Body).Decode(target)
}

func (c *GonicClient) HealthCheck() bool {
	var resp struct {
		SubsonicResponse struct {
			Status string `json:"status"`
		} `json:"subsonic-response"`
	}

	err := c.doRequest("ping", nil, &resp)
	if err != nil {
		return false
	}

	return resp.SubsonicResponse.Status == "ok"
}

// tokenFromPassword generates a Subsonic token = hex(md5(password + salt)).
func tokenFromPassword(password, s string) string {
	h := md5.Sum([]byte(password + s))
	return hex.EncodeToString(h[:])
}

// salt generates a random hex string for Subsonic token-based auth.
func salt() (string, error) {
	b := make([]byte, 4)
	if _, err := rand.Read(b); err != nil {
		return "", fmt.Errorf("salt generation failed: %w", err)
	}
	return hex.EncodeToString(b), nil
}
